WrappedSerialDescriptor.equals doesn't work

**Describe the bug**

Hi, I hit this while creating a generic custom serializer with our format-specific type support. My descriptor is like this

```
override val descriptor: SerialDescriptor = SerialDescriptor("android.util.SparseArray", surrogateSerializer.descriptor)
```

And I can't compare it in `encodeSerializableValue` as the comparison seems to be instance based.

**To Reproduce**
Attach a code snippet or test data if possible.

```
val a = SerialDescriptor("android.util.SparseArray", surrogateSerializer.descriptor)
val b = SerialDescriptor("android.util.SparseArray", surrogateSerializer.descriptor)
assertThat(a).isEqualTo(b)
```

**Expected behavior**

The test should pass, but I got:

```
java.lang.AssertionError: expected:<kotlinx.serialization.descriptors.WrappedSerialDescriptor@a47353e> but was:<kotlinx.serialization.descriptors.WrappedSerialDescriptor@e7ebf9f>
at org.junit.Assert.fail(Assert.java:89)
```

**Environment**
 - Kotlin version: 2.0.20
 - Library version: 1.8.0
 - Kotlin platforms: Android
 - Gradle version: 8.9

--- a/core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptors.kt
+++ b/core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptors.kt
@@ -121,7 +121,25 @@ public fun SerialDescriptor(serialName: String, original: SerialDescriptor): Ser
     return WrappedSerialDescriptor(serialName, original)
 }
 
-internal class WrappedSerialDescriptor(override val serialName: String, original: SerialDescriptor) : SerialDescriptor by original
+internal class WrappedSerialDescriptor(override val serialName: String, private val original: SerialDescriptor) :
+    SerialDescriptor by original {
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is WrappedSerialDescriptor) return false
+
+        return serialName == other.serialName && original == other.original
+    }
+
+    override fun hashCode(): Int {
+        var result = serialName.hashCode()
+        result = 31 * result + original.hashCode()
+        return result
+    }
+
+    override fun toString(): String = toStringImpl()
+
+}
 
 /**
  * An unsafe alternative to [buildClassSerialDescriptor] that supports an arbitrary [SerialKind].
@@ -370,10 +388,5 @@ internal class SerialDescriptorImpl(
 
     override fun hashCode(): Int = _hashCode
 
-    override fun toString(): String {
-        return (0 until elementsCount).joinToString(", ", prefix = "$serialName(", postfix = ")") {
-            getElementName(it) + ": " + getElementDescriptor(it).serialName
-        }
-    }
+    override fun toString(): String = toStringImpl()
 }
-

--- a/core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt
+++ b/core/commonMain/src/kotlinx/serialization/internal/PluginGeneratedSerialDescriptor.kt
@@ -96,11 +96,7 @@ internal open class PluginGeneratedSerialDescriptor(
 
     override fun hashCode(): Int = _hashCode
 
-    override fun toString(): String {
-        return (0 until elementsCount).joinToString(", ", "$serialName(", ")") { i ->
-            getElementName(i) + ": " + getElementDescriptor(i).serialName
-        }
-    }
+    override fun toString(): String = toStringImpl()
 }
 
 @OptIn(ExperimentalSerializationApi::class)
@@ -130,4 +126,8 @@ internal fun SerialDescriptor.hashCodeImpl(typeParams: Array<SerialDescriptor>):
     result = 31 * result + namesHash
     result = 31 * result + kindHash
     return result
+}
+
+internal fun SerialDescriptor.toStringImpl(): String = (0 until elementsCount).joinToString(", ", "$serialName(", ")") { i ->
+    getElementName(i) + ": " + getElementDescriptor(i).serialName
 }

--- a/core/commonTest/src/kotlinx/serialization/WrappedSerialDescriptorTest.kt
+++ b/core/commonTest/src/kotlinx/serialization/WrappedSerialDescriptorTest.kt
@@ -75,4 +75,14 @@ class WrappedSerialDescriptorTest {
         }
         assertFails { SerialDescriptor("kotlin.Int", Int.serializer().descriptor) }
     }
+
+    @Test
+    fun testWrappedEqHc() {
+        val a = SerialDescriptor("SimpleTypeWrapper", SimpleType.serializer().descriptor)
+        val b = SerialDescriptor("SimpleTypeWrapper", SimpleType.serializer().descriptor)
+        assertEquals(a, b)
+        assertEquals(a.hashCode(), b.hashCode())
+        assertNotEquals(a, SimpleType.serializer().descriptor)
+        assertEquals(a.toString(), "SimpleTypeWrapper(int: kotlin.Int, float: kotlin.Float)")
+    }
 }

